llvm/include/llvm/CAS/CASDB.h

```diff
@@ -12,39 +12,362 @@
+///       accessors.
+///     - \a AnyObjectHandle: a variant between subclasses of \a
+///     ObjectHandle.
+///       Returned by \a loadObject().
+///     - \a AnyDataHandle: a variant between NodeHandle (via \a
+///       NodeHandle::getData()) and \a BlobHandle (via \a
+///       BlobHandle::getData()). Returned by \a storeData().
+/// - \a CASID: the UID for an object in the CAS, obtained through \a
+///   CASDB::getObjectID() or \a CASDB::parseID(). This is a valid CAS
+///   identifier, but may reference an object that is unknown to this CAS
+///   instance.
+///
+/// TODO: Once blobs and trees are removed, this picture will be simpler, since
+/// \a ObjectHandle and \a NodeHandle can be collapsed together.
+///
+/// FIXME: Split out ActionCache as a separate concept, and rename this
+/// ObjectStore.
+///
+/// TODO: Consider optimizing small and/or string-like leaf objects:
+///
+/// - \a NodeBuilder and \a NodeReader interfaces can bring some of the same
+///   gains without adding complexity to \a CASDB. E.g., \a NodeBuilder could
+///   have an API to add a named field to a node under construction; if the
+///   name is small enough, it's stored locally in the node's own data, but if
+///   it's bigger then it's outlined to a separate CAS object. \a NodeReader
+///   could handle the complications of reading.
+/// - Implementations can do fast lookups of small objects by adding a
+///   content-based index for them (prefix tree / suffix tree of content),
+///   amortizing overhead of hash computation in \a storeNode().
+/// - Implementations could remove small leaf objects from the main index,
+///   indexing them separately with a partial hash (e.g., 4B prefix), to
+///   optimize storage overhead (32B hash is big for small objects!). Lookups
+///   by UID that miss the main index would get more expensive, requiring a
+///   hash computation for each small object with a matching partial hash, but
+///   maybe this would be rare. To mitigate this cost, small leaf objects could
+///   get added to the main index lazily on first lookup-by-UID, lazily adding
+///   the full overhead of the hash storage only when used by clients.
+/// - NOTE: we tried adding an API to store "raw data" that can be optimized,
+///   but it was very complicated to reason about.
+///     - Introduced many opportunities for implementation bugs.
+///     - Introduced many complications in the API.
+class CASDB : public CASIDContext {
+  void anchor() override;
+
 public:
-  CASID getID() const { return ID; }
-  operator CASID() const { return ID; }
+  /// Get a \p CASID from a \p ObjectRef, which should have been generated by
+  /// \a CASID::print(). This succeeds as long as \p ObjectRef is valid
+  /// (correctly formatted); it does not refer to an object that exists, just
+  /// be a reference that has been constructed correctly.
+  virtual Expected<CASID> parseID(StringRef ObjectRef) = 0;
```

Comment:
Grammar ("just be a").  Maybe "it does not necessarily refer to an object that exists". I don't know what "constructed correctly" means other than what you already said about "correctly formatted".